Add data type support page

neon60 opened the pull request:
Summary of proposed changes:

We are collecting the supported precision support types of ROCm libraries. We have a short summary page, and we create a detailed page at library document.

Short summary page:
https://rocm.docs.amd.com/en/latest/reference/precision-support.html

Summary of proposed changes:

- Add supported data types page
- List the different levels supported types
- Link the reference section on data type support page.

neon60 committed:
Add data type support page

neon60 commented:
#982 Probably this PR should go in first as this is a large change set and also this effect the content of data type support page.

daineAMD commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+Functions data type support
+=======================================
+
+This section colledcted the data tpye support tables of BLAS functions and
```
tpye -> type; colledcted -> collected/collects?

neon60 replied:
Fixed.

daineAMD reviewed (commented):
Before I go through the support for all the functions: the functionality of hipBLAS depends on the backend used, i.e. with AMD/HIP backend, hipBLAS aims at supporting whatever precisions rocBLAS supports, and with CUDA backend, hipBLAS aims at supporting whatever precisions cuBLAS supports. Should we have different sections for each backend? Or should we just guide users to the documentation for the backend? What's being done for other hip* libraries?

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -178,6 +132,14 @@ Graph support (also referred to as stream capture support) for hipBLAS depends o
+Custom data type
+=================
+
+hipBlas defines the ``hipblasBfloat16``, ``hipblasComplex`` and
```
```suggestion
hipBlas defines the ``hipblasBfloat16``, ``hipblasComplex``, and
```

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -178,6 +132,14 @@ Graph support (also referred to as stream capture support) for hipBLAS depends o
+hipBlas defines the ``hipblasBfloat16``, ``hipblasComplex`` and
+``hipblasDoubleComplex`` data types.
+
+For more details, check :ref:`custom_types`.
```
```suggestion
For more details, see :ref:`custom_types`.
```

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -178,6 +132,14 @@ Graph support (also referred to as stream capture support) for hipBLAS depends o
 If rocBLAS is the backend, see the :doc:`rocBLAS <rocblas:index>` documentation.
 Similarly, if CUDA cuBLAS is the backend, see the cuBLAS documentation.
 
+Custom data type
```
```suggestion
Custom data types
```

amd-jnovotny commented on docs/sphinx/_toc.yml.in:
```diff
@@ -26,6 +26,8 @@ subtrees:
 - caption: API Reference
   entries:
+    - file: reference/data-type-support
+      title: Datat type support
```
```suggestion
      title: Data type support
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+Custom types
+============
+
+hipBlas defines the ``hipblasBfloat16``, ``hipblasComplex`` and
```
```suggestion
hipBlas defines the ``hipblasBfloat16``, ``hipblasComplex``, and
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+To expose an API that uses the HIP-defined complex types, use a compiler define
+or inline ``#define ROCM_MATHLIBS_API_USE_HIP_COMPLEX`` before including the
+header file ``<hipblas.h>``. The API is therefore compatible with both forms,
+but recompilation is required to avoid casting if switching to passing in the
```
```suggestion
but recompilation is required to avoid casting if switching to pass in the
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+Functions data type support
+=======================================
+
+This section colledcted the data tpye support tables of BLAS functions and
```
```suggestion
This section collects the data type support tables of BLAS functions and
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+
+        * 
+          - :ref:`DGMM <hipblas_dgmm>`
+          - Diagonal matrix matric multiplication.
```
Do you mean matrix-matrix?

neon60 replied:
Fixed.

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+
+        * 
+          - :ref:`DGMM <hipblas_dgmm>`
+          - Diagonal matrix matric multiplication.
```
Same comment as above

neon60 replied:
Fixed.

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+Extensions
+----------
+
+The extension function data type support is listed for every function separately
```
Is it worth mentioning that these are rocBLAS extensions?

neon60 replied:
I am mentioning here the hipBLAS extensions functions, not rocBLAS.

amd-jnovotny replied:
@neon60 OK, the link goes to the rocBLAS extension page, so that's why I was confused. As long as that's what you want, we're good.

neon60 replied:
Wow, thanks. It's interesting. So the reference should be unique. I pushed the fix.

neon60 replied:
Ctrl+F5 needed for testing

amd-jnovotny replied:
Looks good now!

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+
+        * 
+          - :ref:`GETRS <hipblas_getrs>`
+          - Solve a system of linear equations :math:`AxX=B` after performing an LU factorization using getrf.
```
```suggestion
          - Solve a system of linear equations :math:`AxX=B` after performing an LU factorization using GETRF.
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+
+        * 
+          - :ref:`GETRS <hipblas_getrs>`
+          - Solve a system of linear equations :math:`AxX=B` after performing an LU factorization using getrf.
```
```suggestion
          - Solve a system of linear equations :math:`AxX=B` after performing an LU factorization using GETRF.
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+
+        * 
+          - :ref:`GETRF <hipblas_getrf>`
+          - Computes the LU factorization of a general matrix using partial pivoting with row interchanges.
```
```suggestion
          - Compute the LU factorization of a general matrix using partial pivoting with row interchanges.
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,828 @@
+
+        * 
+          - :ref:`GETRF <hipblas_getrf>`
+          - Computes the LU factorization of a general matrix using partial pivoting with row interchanges.
```
```suggestion
          - Compute the LU factorization of a general matrix using partial pivoting with row interchanges.
```

amd-jnovotny reviewed (approved):
A few typo to fix along with a handful of suggestions.

neon60 commented:
> Before I go through the support for all the functions: the functionality of hipBLAS depends on the backend used, i.e. with AMD/HIP backend, hipBLAS aims at supporting whatever precisions rocBLAS supports, and with CUDA backend, hipBLAS aims at supporting whatever precisions cuBLAS supports. Should we have different sections for each backend? Or should we just guide users to the documentation for the backend? What's being done for other hip* libraries?

At the other hip* libraries, there are various solutions, which is depends on the state of the library:

https://rocm.docs.amd.com/projects/hipRAND/en/latest/api-reference/data-type-support.html
https://rocm.docs.amd.com/projects/hipSPARSELt/en/develop/reference/data-type-support.html
https://rocm.docs.amd.com/projects/hipCUB/en/latest/api-reference/data-type-support.html

If the cuBLAS data types supports are not fully covered, then we should have a detailed list like we have in hipSPARSELt. If it fully supported, we can add just a note. 

I think we should have different sections for different backends. Where can I find information about the different backends in the documentation? Based on the documentation, I only have to differentiate the extensions functions, right?

neon60 commented:
> > > Before I go through the support for all the functions: the functionality of hipBLAS depends on the backend used, i.e. with AMD/HIP backend, hipBLAS aims at supporting whatever precisions rocBLAS supports, and with CUDA backend, hipBLAS aims at supporting whatever precisions cuBLAS supports. Should we have different sections for each backend? Or should we just guide users to the documentation for the backend? What's being done for other hip* libraries?
> > 
> > 
> > @daineAMD , At the other hip* libraries, there are various solutions, which is depends on the state of the library:
> > https://rocm.docs.amd.com/projects/hipRAND/en/latest/api-reference/data-type-support.html https://rocm.docs.amd.com/projects/hipSPARSELt/en/develop/reference/data-type-support.html https://rocm.docs.amd.com/projects/hipCUB/en/latest/api-reference/data-type-support.html
> > If the cuBLAS data types supports are not fully covered, then we should have a detailed list like we have in hipSPARSELt. If it fully supported, we can add just a note.
> > I think we should have different sections for different backends. Where can I find information about the different backends in the documentation? Based on the documentation, I only have to differentiate the extensions functions, right?
> 
> Any datatypes that cuBLAS supports should be supported via hipBLAS as well. There are a couple of functions here and there that are supported in cuBLAS, but not in hipBLAS, but that's kind of aside the point as it's entire functions, not just datatypes.
> 
> However, there are functions available in hipBLAS which will return NOT_SUPPORTED with a cuBLAS backend. I quickly checked the documentation and don't see any list or anything, so I'll put together a list here that you can use for reference.
> 
> * fp16 axpy, dot
> * bf16 dot
> * complex fp32 and complex fp64 spr
> * ALL batched functions except for trsm, gemv, and gemm and solver functions (getrf, getrs, geqrf, gels)
> * ALL strided_batched functions except for gemv and gemm
> * all trtri, all trsm_ex
> * non-batched and strided_batched getrf, getrs, geqrf, gels
> 
> Now, for _ex functions, the best bet is probably to look at the documentation directly: [cublas AxpyEx Doc](https://docs.nvidia.com/cuda/cublas/index.html#cublasaxpyex), below is docs for DotEx, ScalEx, and RotEx.
> 
> [cublas GemmEx Doc](https://docs.nvidia.com/cuda/cublas/index.html#cublasgemmex).
> 
> [rocBLAS axpyEx Doc](https://rocm.docs.amd.com/projects/rocBLAS/en/latest/reference/extension.html#rocblas-axpy-ex-batched-strided-batched), see DotEx, etc. below.
> 
> [rocBLAS gemmEx Doc](https://rocm.docs.amd.com/projects/rocBLAS/en/latest/reference/extension.html#rocblas-gemm-ex-batched-strided-batched)
> 
> I know this is a lot, but thought I'd share as much info as I could and you could decide what to do with it. If you need any clarification please let me know.

Thanks. I need some time to read through and will update the PR accordingly.

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -101,9 +101,9 @@ will be replaced by ``hipDataType``. See the :any:`hipblasGemmEx` documentation
-``hipblasComplex`` and ``hipblasDoubleComplex`` are also deprecated. In a future release,
-all instances of these data types will be replaced with their HIP counterparts
-``hipComplex`` and ``hipDoubleComplex``.
+``hipblasComplex`` and ``hipblasDoubleComplex`` are also deprecated. All 
+instances of these data types replaced with their ``std`` counterparts
+``std::complex<float>`` and ``std::complex<double>``.
```
```suggestion
``hipblasComplex`` and ``hipblasDoubleComplex`` are also deprecated. All 
instances of these data types are replaced with their ``std`` counterparts
``std::complex<float>`` and ``std::complex<double>``.
```

amd-jnovotny reviewed (approved):
Just a small fix I noticed. Let me know if there are more updates I should review based on the SME feedback.

neon60 commented:
> rocBLAS gemmEx Doc

First I added to support page the different data type support for different backends as a tab, but it was not highlighting the differences and only increased the complexity. So I added a section for unsupported cuBLAS backend section, which mentioned on the data type support section.

The extension functions are only linked for the different backends, because we do not want to double document or document 3rd party components.

@amd-jnovotny please review the PR now.

amd-jnovotny commented:
@neon60 I'm not sure which sections I need to review again. Can you clarify?

neon60 commented:
> Just a small fix I noticed. Let me know if there are more updates I should review based on the SME feedback.

These are the changes, which were added after your review:
https://github.com/ROCm/hipBLAS/pull/986/commits/3a8750a3b260bd4cfde11eaf1a25c349256ba3e5

amd-jnovotny commented:
@neon60 Will try to look when I can. Am swamped today.

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,829 @@
+.. note::
+
+  ``hipblasComplex``, ``hipblasDoubleComplex``, and the use of 
+  ``ROCM_MATHLIBS_API_USE_HIP_COMPLEX`` are now deprecated. The API provide
```
```suggestion
  ``ROCM_MATHLIBS_API_USE_HIP_COMPLEX`` are now deprecated. The API provides
```

amd-jnovotny commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,829 @@
+----------
+
+The :ref:`extension functions <hipblas_extension>` data type support is listed
+for every function separately for different backends on the
+:ref:`rocBLAS extensions <rocblas:extension>` or
+`cuBLAS extensions <https://docs.nvidia.com/cuda/cublas/index.html#blas-like-extension>`_ 
+documentation pages.
```
```suggestion
separately for every function for the different backends in the
:ref:`rocBLAS extensions <rocblas:extension>` and
`cuBLAS extensions <https://docs.nvidia.com/cuda/cublas/index.html#blas-like-extension>`_ 
documentation.
```

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -72,6 +72,31 @@ hipBLAS function uses the following notations to denote precisions:
+cuBLAS backend
+==============
+
+hipBLAS has multiple backends for different platforms. cuBLAS for NVIDIA
+platform and rocBLAS for AMD platform. cuBLAS backend does not support all the
+functions and only return ``HIPBLAS_STATUS_NOT_SUPPORTED`` status code.
```
```suggestion
hipBLAS has multiple backends for different platforms: cuBLAS for the NVIDIA
platform and rocBLAS for the AMD platform. The cuBLAS backend does not support all the
functions and only returns the ``HIPBLAS_STATUS_NOT_SUPPORTED`` status code.
```

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -72,6 +72,31 @@ hipBLAS function uses the following notations to denote precisions:
+platform and rocBLAS for AMD platform. cuBLAS backend does not support all the
+functions and only return ``HIPBLAS_STATUS_NOT_SUPPORTED`` status code.
+
+The following level 1-3 and solver funtions are not supported with cuBLAS
```
```suggestion
The following level 1-3 and solver functions are not supported with the cuBLAS
```

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -72,6 +72,31 @@ hipBLAS function uses the following notations to denote precisions:
+  :ref:`hipblas_gels`).
+* All the strided_batched functions except for :ref:`hipblas_gemv` and
+  :ref:`hipblas_gemm`.
+* :ref:`hipblas_trtri` and :ref:`hipblas_trsm_ex` functions.
```
hipblas_trsm_ex link doesn't show up

neon60 replied:
fixed

amd-jnovotny commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -72,6 +72,31 @@ hipBLAS function uses the following notations to denote precisions:
+* All the strided_batched functions except for :ref:`hipblas_gemv` and
+  :ref:`hipblas_gemm`.
+* :ref:`hipblas_trtri` and :ref:`hipblas_trsm_ex` functions.
+* :ref:`hipblas_getrf`, :ref:`hipblas_getrs`, :ref:`hipblas_geqrf` and 
```
```suggestion
* :ref:`hipblas_getrf`, :ref:`hipblas_getrs`, :ref:`hipblas_geqrf`, and 
```

amd-jnovotny reviewed (approved):
Just a handful of comments.

daineAMD commented:
@neon60 #982 has been merged so looks like you'll have some conflicts to go over now, I hope they're minimal. I'll give it another review once resolved.

daineAMD commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,805 @@
+          - ✅
+          - ✅
+
+  .. tab-item:: Complex types
```
May be easier if these are just combined with the above table? Just a suggestion, I'm fine either way.

neon60 replied:
We got a too long table, that's why I had to split. Selected the current form at rocBLAS:
https://rocm.docs.amd.com/projects/rocBLAS/en/develop/reference/data-type-support.html

daineAMD commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,805 @@
+        *
+          - :ref:`GERU and GERC <hipblas_ger>`
+          - Generalized rank-1 update for unconjugated or conjugated complex numbers: :math:`A = \alpha x y^T + A`
+          - ❌
+          - ❌
+          - ❌
+          - ❌
```
N/A might be more fitting here

neon60 replied:
Fixed. I will fix this at rocBLAS side too, after we merged in this PR.

daineAMD commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,805 @@
+        *
+          - :ref:`GER <hipblas_ger>`
+          - Generalized rank-1 update: :math:`A = \alpha x y^T + A`
+          - ❌
+          - ❌
```
Again, maybe N/A

neon60 replied:
Fixed. I will fix this at rocBLAS side too, after we merged in this PR.

daineAMD commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,805 @@
+        * 
+          - :ref:`TRTRI <hipblas_trtri>`
+          - Triangular matrix inversion.
+          - ❌
+          - ❌
```
These should be supported.

neon60 replied:
Fixed.

daineAMD commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,805 @@
+
+        * 
+          - :ref:`GETRI <hipblas_getri>`
+          - Compute the inverse of a matrix using its LU factorization.
```
Only the batched version of getri is supported, not sure if that's noted somewhere.

neon60 replied:
Fixed.

daineAMD commented on docs/reference/hipblas-api-functions.rst:
```diff
@@ -129,6 +136,14 @@ Graph support (also referred to as stream capture support) for hipBLAS depends o
+Custom data types
+=================
+
+hipBlas defines the ``hipblasBfloat16``, ``hipblasComplex``, and
```
hipblasComplex and hipblasDoubleComplex should be removed here as we use hipComplex types now.

neon60 replied:
Fixed.

daineAMD reviewed (approved):
Generally looks good, added some comments that should be addressed.

neon60 committed:
PR feedbacks

daineAMD commented on docs/reference/data-type-support.rst:
```diff
@@ -0,0 +1,809 @@
+
+.. rubric:: Footnotes
+
+.. [#getri] Only the batched GETR functions are supported.
```
typo: GETR -> GETRI

neon60 replied:
Fixed.

neon60 committed:
Fix typo

neon60 commented:
Not blocked, please merge in.
